readthedocs/analytics/proxied_api.py

```diff
@@ -74,12 +75,26 @@ def increase_page_view_count(self, request, project, version, absolute_uri):
             path=path,
             date=timezone.now().date(),
         )
+        if not self._update_view_count(**fields):
+            try:
+                PageView.objects.create(**fields, view_count=1)
+            except IntegrityError:
```

Comment:
Is there any reason to not use `PageView.objects.get_or_create()` instead?

Reply:
get_or_create will fail too https://docs.djangoproject.com/en/3.2/ref/models/querysets/#get-or-create

> concurrent calls to this method may result in multiple rows with the same parameters being inserted.

Reply:
You are not reading the whole note from that page:

> This method is atomic assuming that the database enforces uniqueness of the keyword arguments (see unique or unique_together). If the fields used in the keyword arguments do not have a uniqueness constraint, concurrent calls to this method may result in multiple rows with the same parameters being inserted.

For our use case, the method is atomic. We just need to remove `view_count=1` from the creation to make it initializes at 0 (the default) and immediately after that call `self._update_view_count`.

Reply:
the method is atomic, but will fail when trying to insert a duplicate. And all of our requests are marked as atomic in our settings ;) https://github.com/readthedocs/readthedocs.org/blob/c0d8d5e11439c9b155b651391becfe6d65d35ee5/readthedocs/settings/base.py#L41-L41

Reply:
Hey @stsewd, take a few minutes to try to understand what I'm saying before replying. Also, please take another few minutes to write more elaborated replies.

That said, note that we are using uniqueness constraint in the `fields` passed (`project`, `version`, `path` and `date` --[all of them are marked as `unique_together` in the model](https://github.com/readthedocs/readthedocs.org/blob/6bbaac29527e1e67c1d3ac925225b07097b2a521/readthedocs/analytics/models.py#L41-L42)), that's why I'm saying that the method is atomic for our use case following the documentation that you linked.

Using `get_or_create` does exactly what you have written in the PR, but in a better way. Please, take a look at the source code of it:

https://github.com/django/django/blob/d8c90d4c22cb218f1c170eba086c53d3dff7fbc0/django/db/models/query.py#L573-L596

I think there is no good reason to not use `get_or_create` here. However, if there is one, I'd like you to explain it to me and I want to understand your position.

Reply:
> Hey @stsewd, take a few minutes to try to understand what I'm saying before replying. Also, please take another few minutes to write more elaborated replies.

Please try not to take things personal when replying to a review. A db call being atomic doesn't mean that it won't fail, but as the code you linked it does check for integrity error, so we can use it.